### ISSUE TITLE
[Question] ScienceQA inference questions

### ISSUE THREAD

**Issue by Unrealluver** [haotian-liu/LLaVA#109]
### Question

Greetings! When I run the SQA inference codes. I have a question how the model gets the contexts?

The codes in `.llava/eval/eval_science_qa.py` are:
```
        # conv = default_conversation.copy()
        conv = conv_templates[args.conv_mode].copy()
        conv.append_message(conv.roles[0], qs)
        prompt = conv.get_prompt()
        inputs = tokenizer([prompt])

        input_ids = torch.as_tensor(inputs.input_ids).cuda()

        keywords = ['###']
        stopping_criteria = KeywordsStoppingCriteria(keywords, tokenizer, input_ids)

        with torch.inference_mode():
            output_ids = model.generate(
                input_ids,
                images=images,
                do_sample=True,
                temperature=0.7,
                max_new_tokens=1024,
                stopping_criteria=[stopping_criteria])

        # TODO: new implementation
        input_token_len = input_ids.shape[1]
        n_diff_input_output = (input_ids != output_ids[:, :input_token_len]).sum().item()
        if n_diff_input_output > 0:
            print(f'[Warning] Sample {i}: {n_diff_input_output} output_ids are not the same as the input_ids')
        outputs = tokenizer.batch_decode(output_ids[:, input_token_len:], skip_special_tokens=True)[0]
```

`prompt` is a string, `You are LLaVA, a large language model trained by UW Madison WAIV Lab, based on LLaMA architecture.You are designed to assist human with a variety of tasks using natural language.Follow the instructions carefully.###Human: Hi!###Assistant: Hi there!  How can I help you today?
###Human: Which figure of speech is used in this text?
Sing, O goddess, the anger of Achilles son of Peleus, that brought countless ills upon the Achaeans.
—Homer, The Iliad
Context: N/A
Options: (A) chiasmus (B) apostrophe###`

`outputs` is a string, ` Assistant: LECTURE: Figures of speech are words or phrases that use language in a nonliteral or unusual way. They can make writing more expressive.
Anaphora is the repetition of the same word or words at the beginning of several phrases or clauses.
We are united. We are powerful. We are winners.
Antithesis involves contrasting opposing ideas within a parallel grammatical structure.
I want to help, not to hurt.
Apostrophe is a direct address to an absent person or a nonhuman entity.
Oh, little bird, what makes you sing so beautifully?
Assonance is the repetition of a vowel sound in a series of nearby words.
Try to light the fire.
Chiasmus is an expression in which the second half parallels the first but reverses the order of words.
Never let a fool kiss you or a kiss fool you.
Understatement involves deliberately representing something as less serious or important than it really is.
As you know, it can get a little cold in the Antarctic.
SOLUTION: The text uses apostrophe, a direct address to an absent person or a nonhuman entity.
Sing, O goddess is a direct address to the goddess Pallas Athena.
###`

I noticed that the `gt_ans` of this sample is `{'from': 'gpt', 'value': 'LECTURE: Figures of speech are words or phrases that use language in a nonliteral or unusual way. They can make writing more expressive.\nAnaphora is the repetition of the same word or words at the beginning of several phrases or clauses.\nWe are united. We are powerful. We are winners.\nAntithesis involves contrasting opposing ideas within a parallel grammatical structure.\nI want to help, not to hurt.\nApostrophe is a direct address to an absent person or a nonhuman entity.\nOh, little bird, what makes you sing so beautifully?\nAssonance is the repetition of a vowel sound in a series of nearby words.\nTry to light the fire.\nChiasmus is an expression in which the second half parallels the first but reverses the order of words.\nNever let a fool kiss you or a kiss fool you.\nUnderstatement involves deliberately representing something as less serious or important than it really is.\nAs you know, it can get a little cold in the Antarctic.\nSOLUTION: The text uses apostrophe, a direct address to an absent person or a nonhuman entity.\nO goddess is a direct address to a goddess, a nonhuman entity.\n###\nANSWER: B.'}`

My question is why would the model infer the context that is not seeded to the model? As we can see, except for the last sentence, the model's `output` is totally the same as the `gt_ans`

**Comment by haotian-liu** [comment]
Hi @Unrealluver, thank you for your question and for the close inspection.

I believe this is due to the way ScienceQA is constructed.  If you search for `Figures of speech are words or phrases that use language in a nonliteral or unusual way. They can make writing more expressive` as the keyword in the training set, you'll see this sentence appear multiple times, and this will be followed by different types like `Anaphora`, `Alliteration` etc.

So, the model first recognizes the problem type, and refers to the lecture that it has previously seen, and finally, reasons and generates the solution.

Note that I would say it is almost impossible for us to accidentally mixup the train/test set as the images from train/test/val are in separate folders.

**Comment by Unrealluver** [comment]
Thanks for your reply. Before I open this issue, I have also checked that there is no data leaking. ^_^
